### PR TITLE
Improve logger capability

### PR DESCRIPTION
**Pull request by ethervoid**
- [x] New logger with: plpy, rollbar integration and file
- [x] Added min log level to notify as config option
- [x] Server config to define dataservices environment for: log, third party servers, etc
- [x] Added logger to the SQL functions
- [x] Raise exception on events that should not be logged like reach the quota limit

### CODE REVIEW

**Comment by ethervoid** [comment]
@rafatower take a first look. I've removed the upgrade/downgrade files so should be easier to make a CR

File: server/extension/sql/100_routing_helper.sql [hunk 1]
```diff
@@ -17,19 +17,22 @@ RETURNS cdb_dataservices_server.simple_route AS $$
   from cartodb_services.mapzen.types import polyline_to_linestring
   from cartodb_services.metrics import QuotaService
   from cartodb_services.tools import Coordinate
+  from cartodb_services.tools import Logger
 
   redis_conn = GD["redis_connection_{0}".format(username)]['redis_metrics_connection']
   user_routing_config = GD["user_routing_config_{0}".format(username)]
 
+  logger = Logger(user_routing_config)
```
Comment (rafatower) [inline server/extension/sql/100_routing_helper.sql]: why does the logger need the user routing config to be instantiated?

Comment (ethervoid) [inline server/extension/sql/100_routing_helper.sql]: Because the logger configuration is inside the DB and the config object has all what is needed. I'd love not to do this way but the other way is pass the DB object

File: server/extension/sql/30_admin0.sql [hunk 2]
```diff
@@ -20,12 +22,10 @@ RETURNS Geometry AS $$
...
     except BaseException as e:
-      import sys, traceback
-      type_, value_, traceback_ = sys.exc_info()
+      import sys
       quota_service.increment_failed_service_use()
-      error_msg = 'There was an error trying to geocode using admin0 geocoder: {0}'.format(e)
-      plpy.notice(traceback.format_tb(traceback_))
-      plpy.error(error_msg)
+      logger.error('Error trying to geocode admin0 polygon', sys.exc_info())
+      raise e
```
Comment (rafatower) [inline server/extension/sql/30_admin0.sql]: I'm concerned about confusing users with error messages. At least they were used to see 'There was an error trying to geocode blah blah'

Comment (ethervoid) [inline server/extension/sql/30_admin0.sql]: I was thinking to change the raise e for raise Exception('There was an error trying to...') and avoid more info to avoid leak of sensitive information.

File: server/lib/python/cartodb_services/cartodb_services/mapzen/isolines.py [hunk 3]
```diff
@@ -85,11 +86,7 @@ def calculate_isodistance(self, origin, transport_mode, distance_range):
     def calculate_isoline(self, origin, costing_model, isorange, upper_rmax, cost_variable, unit_factor=1.0):
 
         # NOTE: not for production
-        #logging.basicConfig(level=logging.DEBUG, filename='/tmp/isolines.log')
-        #logging.basicConfig(level=logging.DEBUG)
-        logging.debug('origin = %s' % origin)
-        logging.debug('costing_model = %s' % costing_model)
-        logging.debug('isorange = %d' % isorange)
+        self._logger.debug('Calculate isoline', data={"origin": origin, "costing_model": costing_model, "isorange": isorange})
```
Comment (rafatower) [inline server/lib/python/cartodb_services/cartodb_services/mapzen/isolines.py]: what default log level do you plan for production?

Comment (ethervoid) [inline server/lib/python/cartodb_services/cartodb_services/mapzen/isolines.py]: warning https://github.com/CartoDB/dataservices-api/pull/248/files/ed5cf25e9cb03c9c7d227c4656ef9f5a4c800db7#diff-1c9765409ffd2cf95aa36ab177efe7b7R491

File: server/lib/python/cartodb_services/cartodb_services/metrics/config.py [hunk 4]
```diff
@@ -33,6 +37,23 @@ def username(self):
     def organization(self):
         return self._orgname
 
+    @property
+    def rollbar_api_key(self):
```
Comment (rafatower) [inline server/lib/python/cartodb_services/cartodb_services/metrics/config.py]: this doesn't fit in metrics very well

File: server/lib/python/cartodb_services/cartodb_services/metrics/config.py [hunk 5]
```diff
@@ -162,6 +186,7 @@ def __parse_config(self, filtered_config, db_config):
         if not self._isolines_provider:
             self._isolines_provider = self.DEFAULT_PROVIDER
         self._geocoder_provider = filtered_config[self.GEOCODER_PROVIDER_KEY].lower()
+        self._period_end_date = date_parse(filtered_config[self.PERIOD_END_DATE])
```
Comment (rafatower) [inline server/lib/python/cartodb_services/cartodb_services/metrics/config.py]: why was this changed?

File: server/lib/python/cartodb_services/cartodb_services/metrics/config.py (hunk 4 above)
Comment (ethervoid) [inline server/lib/python/cartodb_services/cartodb_services/metrics/config.py]: I don't think this config is in metrics anymore. Needs refactor but by now I want to add feature not mix refactor and new feature. The two hats syndrome xD

File: server/lib/python/cartodb_services/cartodb_services/metrics/config.py (hunk 5 above)
Comment (ethervoid) [inline server/lib/python/cartodb_services/cartodb_services/metrics/config.py]: I put above. It was below the if, just to clarify. Is not a change

File: server/lib/python/cartodb_services/cartodb_services/metrics/config.py [hunk 6]
```diff
@@ -400,11 +424,23 @@ def __init__(self, db_conn, username, orgname):
...
+        self._get_server_config()
         self._get_here_config()
         self._get_mapzen_config()
         self._get_logger_config()
         self._get_data_observatory_config()
 
+    def _get_server_config(self):
+        server_config_json = self._get_conf('server_conf')
+        if not server_config_json:
+            self._server_environment = 'production'
```
Comment (rafatower) [inline server/lib/python/cartodb_services/cartodb_services/metrics/config.py]: why is `'production'` the default?

File: server/lib/python/cartodb_services/cartodb_services/metrics/config.py [hunk 7]
```diff
@@ -451,6 +487,15 @@ def _get_logger_config(self):
...
             self._geocoder_log_path = logger_conf['geocoder_log_path']
+            self._rollbar_api_key = None
+            self._min_log_level = 'warning'
+            self._log_file_path = None
+            if 'min_log_level' in logger_conf:
+                self._min_log_level = logger_conf['min_log_level']
+            if 'rollbar_api_key' in logger_conf:
+                self._rollbar_api_key = logger_conf['rollbar_api_key']
+            if 'log_file_path' in logger_conf:
+                self._log_file_path = logger_conf['log_file_path']
```
Comment (rafatower) [inline server/lib/python/cartodb_services/cartodb_services/metrics/config.py]: I don't see why this is coupled with here or any other provider

File: server/lib/python/cartodb_services/cartodb_services/metrics/log.py [hunk 8]
```diff
@@ -4,17 +4,17 @@
 import re
 
 
-class LoggerFactory:
+class MetricsLoggerFactory:
```
Comment (rafatower) [inline server/lib/python/cartodb_services/cartodb_services/metrics/log.py]: Here I see I'm missing changes, probably from an earlier PR already merged

File: server/lib/python/cartodb_services/cartodb_services/metrics/config.py (hunk 6 above)
Comment (ethervoid) [inline server/lib/python/cartodb_services/cartodb_services/metrics/config.py]: Because is the default environment right now if the property is missing.

File: server/lib/python/cartodb_services/cartodb_services/metrics/log.py (hunk 8 above)
Comment (rafatower) [inline server/lib/python/cartodb_services/cartodb_services/metrics/log.py]: was this file used before? Seems like it was first committed on March

File: server/lib/python/cartodb_services/cartodb_services/metrics/config.py (hunk 7 above)
Comment (ethervoid) [inline server/lib/python/cartodb_services/cartodb_services/metrics/config.py]: This part is just for logger (metrics and log) is not coupled. The better scenario is have it out of here but I have to create a new function in SQL part to create the config and pass the db connection and so on

File: server/lib/python/cartodb_services/cartodb_services/metrics/log.py (hunk 8 above)
Comment (ethervoid) [inline server/lib/python/cartodb_services/cartodb_services/metrics/log.py]: This is the metrics logger used by systems to gather metrics from the services

Comment (rafatower) [inline server/lib/python/cartodb_services/cartodb_services/metrics/log.py]: Okay, I see we're talking here about different loggers. The resulting files were supposed to be consumed by ES.

File: server/lib/python/cartodb_services/cartodb_services/metrics/config.py (hunk 7 above)
Comment (ethervoid) [inline server/lib/python/cartodb_services/cartodb_services/metrics/config.py]: It's possible to pass the plpy object and create the logger from there...

File: server/lib/python/cartodb_services/cartodb_services/tools/log.py [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+        if data:
+            extra_data['data'] = data
+        else:
+            extra_data['data'] = ''
+
+        return extra_data
+
+    def _setup_log_file_config(self, log_file_path):
+        format_str = "%(asctime)s %(name)-12s %(levelname)-8s %(message)s %(data)s %(exception_type)s %(exception_message)s %(exception_traceback)s"
+        logging.basicConfig(filename=log_file_path, format=format_str, level=self._config.min_log_level.upper())
```
Comment (rafatower) [inline server/lib/python/cartodb_services/cartodb_services/tools/log.py]: I wouldn't use that, it has implications beyond your log module.

I'd instantiate your own logger instead to dump to a file, instead of tweaking `logging` module level config: https://docs.python.org/2.7/howto/logging.html#configuring-logging

File: server/lib/python/cartodb_services/requirements.txt [hunk 10]
```diff
@@ -2,6 +2,7 @@ redis==2.10.5
 hiredis==0.1.5
 python-dateutil==2.2
 googlemaps==2.4.2
+rollbar==0.13.2
```
Comment (rafatower) [inline server/lib/python/cartodb_services/requirements.txt]: Having both `requirements.txt` and a `setup.py` with dependencies is kind of a contradiciton. I would favor the `setup.py`, move all the requirements there and remove the other.

Is this correctly automated with chef or is it done manually?

Comment (ethervoid) [inline server/lib/python/cartodb_services/requirements.txt]: requirements.txt is a default element of python. setup.py is por pip and don't have requirements. This requirements is informative, the package install should be change in chef recipe along with the config part

Comment (rafatower) [inline server/lib/python/cartodb_services/requirements.txt]: of course I might be wrong XD

File: server/lib/python/cartodb_services/cartodb_services/metrics/config.py (hunk 6 above)
Comment (rafatower) [inline server/lib/python/cartodb_services/cartodb_services/metrics/config.py]: why is the default environment right now if the property is missing?

Comment (ethervoid) [inline server/lib/python/cartodb_services/cartodb_services/metrics/config.py]: I've make a change and is development

File: server/extension/sql/100_routing_helper.sql [hunk 11]
```diff
@@ -22,7 +22,7 @@ RETURNS cdb_dataservices_server.simple_route AS $$
   redis_conn = GD["redis_connection_{0}".format(username)]['redis_metrics_connection']
   user_routing_config = GD["user_routing_config_{0}".format(username)]
 
-  logger = Logger(user_routing_config)
+  logger = Logger(plpy)
```
Comment (rafatower) [inline server/extension/sql/100_routing_helper.sql]: what I meant is that it should receive a logger_config object instead, shouldn't it?

File: server/lib/python/cartodb_services/cartodb_services/tools/log.py [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+import logging
+import json
+import traceback
+import sys
+# Monkey path because plpython sys module doesn't have argv and rollbar
+# package use it
+sys.__dict__['argv'] = []
+
+
+class Logger:
```
Comment (rafatower) [inline server/lib/python/cartodb_services/cartodb_services/tools/log.py]: nice to have: are there any tests for this class?

File: server/extension/sql/100_routing_helper.sql (hunk 11 above)
Comment (ethervoid) [inline server/extension/sql/100_routing_helper.sql]: I wouldn't do that because we need to create a new object above the logger that don't provides anything than problems because if it changes in the future we have to change all the SQL. The regular way is to create a config object but in extension i thinks is cumbersome

File: server/lib/python/cartodb_services/cartodb_services/tools/log.py (hunk 12 above)
Comment (ethervoid) [inline server/lib/python/cartodb_services/cartodb_services/tools/log.py]: Nope but I'm thinking if needed because all are third party "services"

File: server/lib/python/cartodb_services/cartodb_services/tools/log.py (hunk 9 above)
Comment (ethervoid) [inline server/lib/python/cartodb_services/cartodb_services/tools/log.py]: I should not have problems tweaking the logging module, the getLogger method is used to difference between modules but we're using one logger. But I'll change it to have a more regular code

File: server/extension/sql/100_routing_helper.sql (hunk 11 above)
Comment (rafatower) [inline server/extension/sql/100_routing_helper.sql]: It is just using a factory for LoggerDBConfig or directly to the Logger if you prefer it. as the code stands now

Comment (ethervoid) [inline server/extension/sql/100_routing_helper.sql]: Changed, now it uses LoggerConfig that is passed in the SQL functions

File: server/lib/python/cartodb_services/cartodb_services/metrics/config.py (hunk 6 above)
Comment (rafatower) [inline server/lib/python/cartodb_services/cartodb_services/metrics/config.py]: I don't understand this.

**Comment by rafatower** [comment]
:+1:

File: server/extension/sql/70_ips.sql [hunk 13]
```diff
@@ -20,12 +23,10 @@ RETURNS Geometry AS $$
...
         return None
     except BaseException as e:
-      import sys, traceback
-      type_, value_, traceback_ = sys.exc_info()
+      import sys
       quota_service.increment_failed_service_use()
-      error_msg = 'There was an error trying to geocode using admin0 geocoder: {0}'.format(e)
-      plpy.notice(traceback.format_tb(traceback_))
-      plpy.error(error_msg)
+      logger.error('Error trying to geocode postal code polygon', sys.exc_info(), data={"username": username, "orgname": orgname})
```
Comment (iriberri) [inline server/extension/sql/70_ips.sql]: This should be IP Address

File: server/extension/sql/80_isolines_helper.sql [hunk 14]
```diff
@@ -44,12 +48,10 @@ RETURNS SETOF cdb_dataservices_server.isoline AS $$
...
       return []
   except BaseException as e:
-    import sys, traceback
-    type_, value_, traceback_ = sys.exc_info()
+    import sys
     quota_service.increment_failed_service_use()
-    error_msg = 'There was an error trying to obtain isodistances using here maps geocoder: {0}'.format(e)
-    plpy.notice(traceback.format_tb(traceback_))
-    plpy.error(error_msg)
+    logger.error('Error trying to get mapzen isolines', sys.exc_info(), data={"username": username, "orgname": orgname})
```
Comment (iriberri) [inline server/extension/sql/80_isolines_helper.sql]: The client here is HERE, isn't it?

File: server/lib/python/cartodb_services/cartodb_services/mapzen/matrix_client.py [hunk 15]
```diff
@@ -40,6 +42,16 @@ def one_to_many(self, locations, costing):
         }
         response = requests.get(self.ONE_TO_MANY_URL, params=request_params)
 
-        response.raise_for_status() # raise exception if not 200 OK
+        raise Exception('testing exception handling')
```
Comment (iriberri) [inline server/lib/python/cartodb_services/cartodb_services/mapzen/matrix_client.py]: ?

File: server/lib/python/cartodb_services/cartodb_services/mapzen/routing.py [hunk 16]
```diff
@@ -48,7 +49,11 @@ def calculate_route_point_to_point(self, waypoints, mode,
         elif response.status_code == requests.codes.bad_request:
             return MapzenRoutingResponse(None, None, None)
         else:
-            response.raise_for_status()
+            self._logger.error('Error trying to calculate route using HERE',
```
Comment (iriberri) [inline server/lib/python/cartodb_services/cartodb_services/mapzen/routing.py]: It's Mapzen routing
